{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7ff2c0b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting face-recognition\n",
      "  Using cached face_recognition-1.3.0-py2.py3-none-any.whl (15 kB)\n",
      "Collecting face-recognition-models>=0.3.0 (from face-recognition)\n",
      "  Using cached face_recognition_models-0.3.0.tar.gz (100.1 MB)\n",
      "  Preparing metadata (setup.py): started\n",
      "  Preparing metadata (setup.py): finished with status 'done'\n",
      "Requirement already satisfied: Click>=6.0 in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (from face-recognition) (8.1.7)\n",
      "Collecting dlib>=19.7 (from face-recognition)\n",
      "  Downloading dlib-19.24.2.tar.gz (11.8 MB)\n",
      "     ---------------------------------------- 0.0/11.8 MB ? eta -:--:--\n",
      "     ---------------------------------------- 0.0/11.8 MB ? eta -:--:--\n",
      "     --------------------------------------- 0.0/11.8 MB 262.6 kB/s eta 0:00:45\n",
      "     --------------------------------------- 0.1/11.8 MB 469.7 kB/s eta 0:00:25\n",
      "      --------------------------------------- 0.2/11.8 MB 1.1 MB/s eta 0:00:11\n",
      "     - -------------------------------------- 0.4/11.8 MB 1.6 MB/s eta 0:00:07\n",
      "     - -------------------------------------- 0.5/11.8 MB 2.0 MB/s eta 0:00:06\n",
      "     -- ------------------------------------- 0.7/11.8 MB 2.4 MB/s eta 0:00:05\n",
      "     --- ------------------------------------ 1.0/11.8 MB 2.7 MB/s eta 0:00:05\n",
      "     --- ------------------------------------ 1.2/11.8 MB 2.9 MB/s eta 0:00:04\n",
      "     ----- ---------------------------------- 1.5/11.8 MB 3.3 MB/s eta 0:00:04\n",
      "     ------ --------------------------------- 1.8/11.8 MB 3.7 MB/s eta 0:00:03\n",
      "     ------- -------------------------------- 2.1/11.8 MB 3.8 MB/s eta 0:00:03\n",
      "     -------- ------------------------------- 2.4/11.8 MB 4.1 MB/s eta 0:00:03\n",
      "     --------- ------------------------------ 2.9/11.8 MB 4.5 MB/s eta 0:00:02\n",
      "     ----------- ---------------------------- 3.3/11.8 MB 4.8 MB/s eta 0:00:02\n",
      "     ------------ --------------------------- 3.8/11.8 MB 5.1 MB/s eta 0:00:02\n",
      "     -------------- ------------------------- 4.1/11.8 MB 5.3 MB/s eta 0:00:02\n",
      "     --------------- ------------------------ 4.4/11.8 MB 5.3 MB/s eta 0:00:02\n",
      "     --------------- ------------------------ 4.7/11.8 MB 5.4 MB/s eta 0:00:02\n",
      "     ----------------- ---------------------- 5.0/11.8 MB 5.4 MB/s eta 0:00:02\n",
      "     ------------------ --------------------- 5.3/11.8 MB 5.5 MB/s eta 0:00:02\n",
      "     ------------------- -------------------- 5.6/11.8 MB 5.5 MB/s eta 0:00:02\n",
      "     -------------------- ------------------- 5.9/11.8 MB 5.5 MB/s eta 0:00:02\n",
      "     -------------------- ------------------- 6.2/11.8 MB 5.6 MB/s eta 0:00:02\n",
      "     --------------------- ------------------ 6.5/11.8 MB 5.6 MB/s eta 0:00:01\n",
      "     ---------------------- ----------------- 6.8/11.8 MB 5.6 MB/s eta 0:00:01\n",
      "     ----------------------- ---------------- 7.0/11.8 MB 5.6 MB/s eta 0:00:01\n",
      "     ------------------------ --------------- 7.3/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     ------------------------- -------------- 7.6/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     -------------------------- ------------- 7.8/11.8 MB 5.6 MB/s eta 0:00:01\n",
      "     --------------------------- ------------ 8.2/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     ---------------------------- ----------- 8.5/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     ----------------------------- ---------- 8.8/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     ------------------------------ --------- 9.1/11.8 MB 5.7 MB/s eta 0:00:01\n",
      "     ------------------------------- -------- 9.4/11.8 MB 5.8 MB/s eta 0:00:01\n",
      "     -------------------------------- ------- 9.6/11.8 MB 5.8 MB/s eta 0:00:01\n",
      "     --------------------------------- ------ 9.9/11.8 MB 5.8 MB/s eta 0:00:01\n",
      "     ---------------------------------- ----- 10.2/11.8 MB 5.8 MB/s eta 0:00:01\n",
      "     ----------------------------------- ---- 10.5/11.8 MB 6.3 MB/s eta 0:00:01\n",
      "     ------------------------------------ --- 10.8/11.8 MB 6.4 MB/s eta 0:00:01\n",
      "     ------------------------------------- -- 11.1/11.8 MB 6.5 MB/s eta 0:00:01\n",
      "     -------------------------------------- - 11.4/11.8 MB 6.6 MB/s eta 0:00:01\n",
      "     ---------------------------------------  11.6/11.8 MB 6.5 MB/s eta 0:00:01\n",
      "     ---------------------------------------  11.8/11.8 MB 6.5 MB/s eta 0:00:01\n",
      "     ---------------------------------------- 11.8/11.8 MB 6.4 MB/s eta 0:00:00\n",
      "  Installing build dependencies: started\n",
      "  Installing build dependencies: finished with status 'done'\n",
      "  Getting requirements to build wheel: started\n",
      "  Getting requirements to build wheel: finished with status 'done'\n",
      "  Preparing metadata (pyproject.toml): started\n",
      "  Preparing metadata (pyproject.toml): finished with status 'done'\n",
      "Requirement already satisfied: numpy in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (from face-recognition) (1.26.2)\n",
      "Requirement already satisfied: Pillow in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (from face-recognition) (9.5.0)\n",
      "Requirement already satisfied: colorama in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (from Click>=6.0->face-recognition) (0.4.6)\n",
      "Building wheels for collected packages: dlib, face-recognition-models\n",
      "  Building wheel for dlib (pyproject.toml): started\n",
      "  Building wheel for dlib (pyproject.toml): finished with status 'error'\n",
      "  Building wheel for face-recognition-models (setup.py): started\n",
      "  Building wheel for face-recognition-models (setup.py): finished with status 'done'\n",
      "  Created wheel for face-recognition-models: filename=face_recognition_models-0.3.0-py2.py3-none-any.whl size=100566184 sha256=a29e8433e77b13d8864ef9508c09f9003b7579d6c2be49ab9e8a800957e44ec4\n",
      "  Stored in directory: c:\\users\\user-pc\\appdata\\local\\pip\\cache\\wheels\\04\\52\\ec\\9355da79c29f160b038a20c784db2803c2f9fa2c8a462c176a\n",
      "Successfully built face-recognition-models\n",
      "Failed to build dlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  error: subprocess-exited-with-error\n",
      "  \n",
      "  Building wheel for dlib (pyproject.toml) did not run successfully.\n",
      "  exit code: 1\n",
      "  \n",
      "  [78 lines of output]\n",
      "  running bdist_wheel\n",
      "  running build\n",
      "  running build_ext\n",
      "  <string>:125: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.\n",
      "  Building extension for Python 3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)]\n",
      "  Invoking CMake setup: 'cmake C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-install-6ae5owjj\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\tools\\python -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-install-6ae5owjj\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\build\\lib.win-amd64-cpython-311 -DPYTHON_EXECUTABLE=C:\\Users\\USER-PC\\anaconda3\\python.exe -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-install-6ae5owjj\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\build\\lib.win-amd64-cpython-311 -A x64'\n",
      "  -- Building for: NMake Makefiles\n",
      "  CMake Error at CMakeLists.txt:5 (message):\n",
      "  \n",
      "  \n",
      "  \n",
      "    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "  \n",
      "  \n",
      "    You must use Visual Studio to build a python extension on windows.  If you\n",
      "    are getting this error it means you have not installed Visual C++.  Note\n",
      "    that there are many flavors of Visual Studio, like Visual Studio for C#\n",
      "    development.  You need to install Visual Studio for C++.\n",
      "  \n",
      "  \n",
      "    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",
      "  \n",
      "  \n",
      "  \n",
      "  \n",
      "  -- Configuring incomplete, errors occurred!\n",
      "  Traceback (most recent call last):\n",
      "    File \"C:\\Users\\USER-PC\\anaconda3\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 353, in <module>\n",
      "      main()\n",
      "    File \"C:\\Users\\USER-PC\\anaconda3\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 335, in main\n",
      "      json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\USER-PC\\anaconda3\\Lib\\site-packages\\pip\\_vendor\\pyproject_hooks\\_in_process\\_in_process.py\", line 251, in build_wheel\n",
      "      return _build_backend().build_wheel(wheel_directory, config_settings,\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 404, in build_wheel\n",
      "      return self._build_with_temp_dir(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 389, in _build_with_temp_dir\n",
      "      self.run_setup()\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\build_meta.py\", line 311, in run_setup\n",
      "      exec(code, locals())\n",
      "    File \"<string>\", line 218, in <module>\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\__init__.py\", line 103, in setup\n",
      "      return distutils.core.setup(**attrs)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\core.py\", line 185, in setup\n",
      "      return run_commands(dist)\n",
      "             ^^^^^^^^^^^^^^^^^^\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\core.py\", line 201, in run_commands\n",
      "      dist.run_commands()\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 969, in run_commands\n",
      "      self.run_command(cmd)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "      super().run_command(command)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "      cmd_obj.run()\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\wheel\\bdist_wheel.py\", line 368, in run\n",
      "      self.run_command(\"build\")\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\cmd.py\", line 318, in run_command\n",
      "      self.distribution.run_command(command)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "      super().run_command(command)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "      cmd_obj.run()\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\command\\build.py\", line 131, in run\n",
      "      self.run_command(cmd_name)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\cmd.py\", line 318, in run_command\n",
      "      self.distribution.run_command(command)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\dist.py\", line 963, in run_command\n",
      "      super().run_command(command)\n",
      "    File \"C:\\Users\\USER-PC\\AppData\\Local\\Temp\\pip-build-env-gw5k72du\\overlay\\Lib\\site-packages\\setuptools\\_distutils\\dist.py\", line 988, in run_command\n",
      "      cmd_obj.run()\n",
      "    File \"<string>\", line 130, in run\n",
      "    File \"<string>\", line 167, in build_extension\n",
      "    File \"C:\\Users\\USER-PC\\anaconda3\\Lib\\subprocess.py\", line 413, in check_call\n",
      "      raise CalledProcessError(retcode, cmd)\n",
      "  subprocess.CalledProcessError: Command '['cmake', 'C:\\\\Users\\\\USER-PC\\\\AppData\\\\Local\\\\Temp\\\\pip-install-6ae5owjj\\\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\\\tools\\\\python', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=C:\\\\Users\\\\USER-PC\\\\AppData\\\\Local\\\\Temp\\\\pip-install-6ae5owjj\\\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\\\build\\\\lib.win-amd64-cpython-311', '-DPYTHON_EXECUTABLE=C:\\\\Users\\\\USER-PC\\\\anaconda3\\\\python.exe', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE=C:\\\\Users\\\\USER-PC\\\\AppData\\\\Local\\\\Temp\\\\pip-install-6ae5owjj\\\\dlib_5fc1bea1d3724ae38288853cb8b485cf\\\\build\\\\lib.win-amd64-cpython-311', '-A', 'x64']' returned non-zero exit status 1.\n",
      "  [end of output]\n",
      "  \n",
      "  note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  ERROR: Failed building wheel for dlib\n",
      "ERROR: Could not build wheels for dlib, which is required to install pyproject.toml-based projects\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting cmake\n",
      "  Obtaining dependency information for cmake from https://files.pythonhosted.org/packages/1e/82/da0c325e829b9987df30d0252851b68ff2b92ad6ffa675817354abe015ad/cmake-3.28.3-py2.py3-none-win_amd64.whl.metadata\n",
      "  Using cached cmake-3.28.3-py2.py3-none-win_amd64.whl.metadata (6.5 kB)\n",
      "Using cached cmake-3.28.3-py2.py3-none-win_amd64.whl (35.7 MB)\n",
      "Installing collected packages: cmake\n",
      "Successfully installed cmake-3.28.3\n",
      "Requirement already satisfied: opencv-python in c:\\users\\user-pc\\anaconda3\\lib\\site-packages (4.9.0.80)\n",
      "Requirement already satisfied: numpy>=1.21.2 in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (from opencv-python) (1.26.2)\n",
      "Requirement already satisfied: numpy in c:\\users\\user-pc\\appdata\\roaming\\python\\python311\\site-packages (1.26.2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Could not find a version that satisfies the requirement as (from versions: none)\n",
      "ERROR: No matching distribution found for as\n"
     ]
    }
   ],
   "source": [
    "# Facial Recognition App:-\n",
    "\n",
    "! pip install face-recognition\n",
    "! pip install cmake\n",
    "! pip install opencv-python\n",
    "! pip install numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4937848e",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'face_recognition'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mface_recognition\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mcv2\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'face_recognition'"
     ]
    }
   ],
   "source": [
    "import face_recognition\n",
    "import cv2\n",
    "import numpy as np\n",
    "import csv\n",
    "from datetime import datetime\n",
    "\n",
    "video_capture =cv2.VideoCapture(0)\n",
    "\n",
    "# Load known faces:\n",
    "akash_image = face_recognition.load_image_file(\"faces/akash.jpg\")\n",
    "akash_encoding = face_recognition.face_encodings(harrys_image)[0]\n",
    "\n",
    "parth_image = face_recognition.load_image_file(\"faces/parth.jpg\")\n",
    "parth_encoding = face_recognition.face_encodings(parth_image)[0]\n",
    "\n",
    "aanya_image = face_recognition.load_image_file(\"faces/aanya.jpg\")\n",
    "aanya_encoding = face_recognition.face_encodings(aanya_image)[0]\n",
    "\n",
    "\n",
    "known_face_encodings = [akash_encoding, parth_encoding, aanya_encoding]\n",
    "known_face_names = [\"Akash\",\"Parth\", \"Aanaya\"]\n",
    "\n",
    "# List of family members:\n",
    "family_members = known_face_names.copy()\n",
    "\n",
    "face_locations = []\n",
    "face_encodings = []\n",
    "\n",
    "# Get the current Date and Time:\n",
    "now = datetime.now()\n",
    "current_date = datetime.strftime(\"%Y-%m-%d\")\n",
    "\n",
    "f = open(f\"{current_date}.csv\", \"w+\", newline=\"\")\n",
    "lnwriter = csv.writer(f)\n",
    "\n",
    "while True:\n",
    "    _, frame = video_capture.read()\n",
    "    small_frame = cv2.resize(frame, (0,0), fx=0.25, fy=0.25)\n",
    "    rgbsmall_frame = cv2.cvtColor(small_frame, cv2.COLOR_BGR2RGB)\n",
    "    \n",
    "    #Recognizes Faces\n",
    "    face_locations = face_recognition.face_locations(rgb_small_frame)\n",
    "    face_encodings = face_recognition.face_encodings(rgb_small_frame, face_locations)\n",
    "    \n",
    "    for face_encodings in face_encodings:\n",
    "        matches = face_recognition.compare_faces(known_face_encodings, face_encoding)\n",
    "        face_distance = face_recognition.face_distance(known_face_encodings,face_encoding)\n",
    "        best_match_index = np.argmin(face_distance)\n",
    "        \n",
    "        if (matches[best_matcch_index]):\n",
    "            name = known_face_names[best_match_index]\n",
    "            \n",
    "        # Add the text if the person is present:\n",
    "        if name in know_face_names:\n",
    "            font = cv2.FONT_HERSHEY_SIMPLEX\n",
    "            bottomLeftCornerOfText = (10, 100)\n",
    "            fontScale = 1.5\n",
    "            fontColor = (255, 0, 0)\n",
    "            thickness = 3\n",
    "            lineType = 2\n",
    "            cv2.putText(frame, name + \"Present\", bottomLeftCornerOfText, font, fontScale, fontColor, thickness, linkType)\n",
    "            \n",
    "            if name in family_members:\n",
    "                family_members.remove(name)\n",
    "                current_time = now.strftime(\"%H-%M%S\")\n",
    "                lnwriter.writerow([name, current_time])\n",
    "            \n",
    "    cv2.imshow(\"Cheese\", frame)\n",
    "    if cv2.waitkey(1) & 0xff == ord(\"q\"):\n",
    "        break\n",
    "\n",
    "videos_capture.release()\n",
    "cv2.destroyAllWindows()\n",
    "f.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
